internal/pkg/agent/application/info/agent_info.go

```diff
@@ -28,10 +29,13 @@ func NewAgentInfoWithLog(level string, createAgentID bool) (*AgentInfo, error) {
 		return nil, err
 	}
 
+	log := logger.NewWithoutConfig("agent_info")
```

Comment:
`func NewAgentInfoWithLog(level string`

Doesn't this ignore the log level passed to this function?

Reply:
It does.  For some bizarre reason, if I try call [`logger. NewWithLogpLevel `](https://github.com/elastic/elastic-agent/blob/2ca71d09b4103c5bf634018912d6d60aad747b52/pkg/core/logger/logger.go#L50) or even [`logger.New`](https://github.com/elastic/elastic-agent/blob/2ca71d09b4103c5bf634018912d6d60aad747b52/pkg/core/logger/logger.go#L45), the Beats logs stop appearing in the Agent log.   I'm still trying to figure out why this is happening but my workaround for the time being was to call `logger.NewWithoutConfig`.

Reply:
What's interested is that, when using `logger.NewWithLogpLevel`, every so often I'm able to get logs with `log.source == "system/metrics-default"` to show up in Agent logs.  This suggests to me that the code paths taken by both [logger.NewWithLogpLevel ](https://github.com/elastic/elastic-agent/blob/2ca71d09b4103c5bf634018912d6d60aad747b52/pkg/core/logger/logger.go#L50) and [logger.New](https://github.com/elastic/elastic-agent/blob/2ca71d09b4103c5bf634018912d6d60aad747b52/pkg/core/logger/logger.go#L45) end up altering some state that comes into play when the Command Runtime tries to redirect stdout/stderr to the Agent's logger.

Reply:
I know where the shared state is coming from: https://github.com/elastic/elastic-agent-libs/blob/ef76bc6e4d942018de95fb6de731ee0d6258a160/logp/core.go#L44

But I'm no closer to figuring out how to make things work here so we're using the log level that's passed in here AND making sure Beats logs show up as part of Agent logs. 😕 

And simply using `log := logger.NewWithoutConfig("agent_info")` isn't going to work because the logger that returns is configured at `info` level and the logging we want when FQDN lookup fails is at the `debug` level.

Reply:
Tangentially related, while working on this problem I noticed that with a Fleet-managed Agent, when you change the log level in the Fleet UI for the Agent in question to `debug`, messages such as [this one](https://github.com/elastic/elastic-agent/blob/2ca71d09b4103c5bf634018912d6d60aad747b52/internal/pkg/agent/application/gateway/fleet/fleet_gateway.go#L157) don't appear in the Agent's logs, both in Discover and the ones obtained via `elastic-agent diagnostics`.  And this is happening with Agent built off `main`, not this PR.

At this point, I'm going to need help from @blakerouse or @cmacknz to figure out how to make logging work as expected so we can log, at the debug level, when FQDN lookup fails over [here](https://github.com/AndersonQ/elastic-agent/blob/ecabe78006ae5efa8a5f0ed35164ec4c63631d3d/internal/pkg/agent/application/info/agent_metadata.go#L150).

Reply:
you're persisting this logger for 2 debug logs in ECSMetadata
wouldn't it be better to inject logger into ECSMetadata call? this is called from places where logger is properly initialized. you can create named logger out of that `c.log.Named("agent_info").Debug("")` (if you really want to)

Reply:
I think setting the agent log level through the configuration might actually have been broken before this change https://github.com/elastic/elastic-agent/issues/2399

Reply:
If this was broken before this change what should be the course of action? Should we move forward with this PR and have a follow up issue to fix the logging problem?

Reply:
@michalpristas Thanks for the suggestion.  I tried it (via 31872110ad80ccfc52cd25e13aa3c5b8c45023e6) and then tested it with Fleet-managed Agent.  It still didn't work - screenshots and logs below:

Debug logs are coming in to Discover after changing the log level in the UI to `debug`:
<img width="1904" alt="Screenshot 2023-04-04 at 06 55 23" src="https://user-images.githubusercontent.com/51061/229816977-2492a81a-bd57-4718-b3a7-4378980f37a9.png">

But no debug logs matching the string I put in one of the log messages in 31872110ad80ccfc52cd25e13aa3c5b8c45023e6:

<img width="1904" alt="Screenshot 2023-04-04 at 06 55 27" src="https://user-images.githubusercontent.com/51061/229817239-a36d0a64-1180-40b0-ae17-a62777a93637.png">

Same with logs produced by `elastic-agent diagnostics`:

```
$ sudo elastic-agent diagnostics
[WARNING] Could not redact state.yaml due to unmarshalling error: yaml: invalid map key: map[interface {}]interface {}{"unitid":"log-default-logfile-system-72201dc0-d2be-11ed-aa4e-19380f49b3e2", "unittype":0}
Created diagnostics archive "elastic-agent-diagnostics-2023-04-04T13-57-16Z-00.zip"
***** WARNING *****
Created archive may contain plain text credentials.
Ensure that files in archive are redacted before sharing.
$ unzip elastic-agent-diagnostics-*.zip -d diag/
... unzip output ...
$ cd diag
$  rg 'ECSMetadata' ./logs
```

I do see Beats logs in Agent logs with this approach, though:

```
$ rg '"log":\{"source":"elastic-agent"\}' --invert-match ./logs
./logs/elastic-agent-20230404.ndjson:{"log.level":"info","@timestamp":"2023-04-04T13:53:12.131Z","message":"Home path: [/opt/Elastic/Agent/data/elastic-agent-55972b/components] Config path: [/opt/Elastic/Agent/data/elastic-agent-55972b/components] Data path: [/opt/Elastic/Agent/data/elastic-agent-55972b/run/system/metrics-default] Logs path: [/opt/Elastic/Agent/data/elastic-agent-55972b/components/logs]","component":{"binary":"metricbeat","dataset":"elastic_agent.metricbeat","id":"system/metrics-default","type":"system/metrics"},"log":{"source":"system/metrics-default"},"ecs.version":"1.6.0","log.origin":{"file.line":785,"file.name":"instance/beat.go"},"service.name":"metricbeat","ecs.version":"1.6.0"}
... lots more output...
```

So I think the approach of injecting the logger is at least _safer_, in that it doesn't break Beats logs showing up in Agent logs.  So I think we should keep the changes in 31872110ad80ccfc52cd25e13aa3c5b8c45023e6. 

But I do think that changing log level as a whole isn't working as expected - whether through standalone configuration or Fleet-managed configuration.  @cmacknz  Given that this is a known issue, even prior to this PR, can we move forward with the changes in this PR?

Reply:
There are currently no reported issues when changing the log level from Fleet. We need to confirm we did not introduce a regression there.

There is a known issue when changing the log level in the agent configuration file. This is a completely separate code path from when the log level is changed in Fleet.

I don't think we should block this PR on issues that also exist on main, we just need to be careful that we are not making things worse before merging.

Reply:
> There are currently no reported issues when changing the log level from Fleet. We need to confirm we did not introduce a regression there.

The issue I found with changing log level from Fleet was with `main`.  I mentioned it earlier in this thread: https://github.com/elastic/elastic-agent/pull/2218#discussion_r1156881592.  Maybe, as a next step, someone other than me should try to reproduce this to prove that it's not a regression from this PR?

Reply:
> I'll make a separate issue for this bug with more details.

https://github.com/elastic/elastic-agent/issues/2450